### DISCUSSION TITLE
Configuring I2S chip

### DISCUSSION THREAD

**Discussion by johanneswilkens** [pschatzmann/arduino-audio-tools#647] [Q&A] [answered]
Hi, I'm just starting with ESP32 audio. I'm struggling a bit with the setup of the audio chip. I want to (eventually) have whole house multiroom audio. I tried SqueezeliteESP32 which seems to work fine. Now I would like to see if I can convert the line signal from a phono preamp to an audio stream which can be opened in LMS.
For development I have the AI thinker development board with the A1S with the following pinout (copied from Squeezelite config): `model=ES8388,bck=27,ws=25,do=26,sda=33,scl=32,i2c=16`

Now when trying to write the code for an audio stream I see there is also a audio kit library, do I need to use that or will the following example code also work?

```
#include <Arduino.h>
#include "AudioTools.h"

AudioWAVServer server("MyWiFiSSID","MyWiFiPass"); 

I2SStream i2sStream;
ConverterFillLeftAndRight<int16_t> filler(LeftIsEmpty);

void setup(){
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.i2s_format = I2S_STD_FORMAT; 
  config.sample_rate = 44100;
  config.channels = 2;
  config.bits_per_sample = 32;
  // i2s pins used on ESP32
  config.pin_bck = 27;
  config.pin_ws = 25;
  config.pin_data = 26;
  i2sStream.begin(config);

  server.begin(i2sStream, config, &filler);
}

void loop() {
  server.copy();  
}
```

Because I'm completely new to this topic and I2S I don't fully understand the difference between all kinds of chips and why this particular chip also has a lot of config options via I2C. I don't see any config for I2C in this code, but I also don't see I2C options in the audiokit library, so I'm a bit confused.

Eventually I'll make my own universal PCB with WROOM module and I2S DAC/ADC (which can either be configured as analog audio to network stream or as network (stream, Spotify, media server, etc.) to audio), probably without Chinese ES8388, but with something from Texas Instruments, so I would like to be able to just change some pins in code and have the same code running on my own board.

**Comment by pschatzmann** [comment, marked as answer]
The I2SStream is only good enough if you use a DAC or ADC that does not need any software configuration. I suggest that you go thru the Hardware Wiki of the AudioTools to see the different options.

For audio chips that require a separate software driver the architecture of the AudioTools is to provide separate Stream classes: in your case this is the AudioKitStream. These classes all rely on separate Driver projects. In your case the AudioKit project: https://github.com/pschatzmann/arduino-audiokit

How to set up your AudioKit is described in the Readme: : If I got it right, your described pins match with option 5 which is set as default. 

I suggest that you also read the documentation in the Wiki of the AudioKit to get more information about your board.

**Comment by johanneswilkens** [comment]
Thanks a lot for your quick reply!

So if I understand correctly, chips without I2C don't need any extra library and can just be hooked up to the I2S pins of the ESP and use the I2SStream class, chips that do have I2C/SPI communication require an extra library for the setup of the chip, and use the class from the library, otherwise it won't work. 

Of course I'm also reading the wiki, but it wasn't entirely clear to me if the extra library was mandatory or optional.

**Comment by johanneswilkens** [comment]
It's working by trial and error I selected `config.input_device = AUDIO_HAL_ADC_INPUT_LINE2;` which is the line input of the Ai Thinker board, only problem is that the signal seems to be interpreted as mic input. The audio is clipped, is there somewhere a wiki page about all possible config settings for the I2C chips?
```
typedef enum {
    AUDIO_HAL_CODEC_MODE_ENCODE = 1,  /*!< select adc */
    AUDIO_HAL_CODEC_MODE_DECODE,      /*!< select dac */
    AUDIO_HAL_CODEC_MODE_BOTH,        /*!< select both adc and dac */
    AUDIO_HAL_CODEC_MODE_LINE_IN,     /*!< set adc channel */
} audio_hal_codec_mode_t;
```
My guess would be that this should be set to Line In mode, but where can I do this in the library?

**Reply by pschatzmann** [reply]
Just look at the AudioKit examples in this project! You define the input or output mode when you request the default configuration: RX_MODE is for getting only input, TX_MODE is only supporting output and RX_TXMODE is doing both.

E.g this [example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-serial/streams-audiokit-serial.ino) prints the input from the microphone to the serial printer

**Comment by pschatzmann** [comment]
This is a known bug in the es8388 based AI Thinker boards: You can't deactivate the microphones...

**Comment by johanneswilkens** [comment]
Oh that’s quite sad. So the best solution to remove the clipping for now would be to add a voltage divider before the board line input to convert line signal from the record player to a mic signal?

**Comment by pschatzmann** [comment]
When I am connecting devices via AUX and put up the volume of the input device, the input is usually so high that the microphone is not disturbing. 

The recommended solution if you have only a weak input signal is to desolder the microphones. 

ps. the AudioKit with the AC101 and LyraT boards (which also have a es8388) do not have this problem.

**Comment by johanneswilkens** [comment]
The mics are not really a problem now, the issue is that it seems to expect a mic level, so I assume it is increasing the gain with about +50dB, but I’m already giving it a line level so it’s clipping on the web stream.

**Comment by pschatzmann** [comment]
So why don't you just change the input volume / gain ?
see https://github.com/pschatzmann/arduino-audiokit/discussions/62
I also updated the FAQ in the AudioKit.

**Reply by johanneswilkens** [reply]
I thought maybe it’s possible to disable the gain in the chip, otherwise I have to cut the cable or solder a different connector to the pcb and put a resistor divider or log potmeter in between. 

Because it’s all DIY stuff I have the volume knob on the amplifier, so it’s not just turning the volume down.
